module/fork_spec.go

```diff
@@ -8,29 +17,229 @@ const (
+			intermediates = append(intermediates, mid)
+		}
+	}
+	for mid := prevLast + prevForkSpec.EpochLength; mid < currentFirst; mid += prevForkSpec.EpochLength {
+		intermediates = append(intermediates, mid)
+	}
+
+	return &BoundaryEpochs{
+		PreviousForkSpec: prevForkSpec,
+		CurrentForkSpec:  b.CurrentForkSpec,
+		BoundaryHeight:   boundaryHeight,
+		PrevLast:         prevLast,
+		CurrentFirst:     currentFirst,
+		Intermediates:    intermediates,
+	}, nil
+}
+
+func (be BoundaryEpochs) CurrentEpochBlockNumber(number uint64) uint64 {
+	if number >= be.CurrentFirst {
+		return number - (number % be.CurrentForkSpec.EpochLength)
+	}
+
+	if len(be.Intermediates) > 0 {
+		for i := len(be.Intermediates) - 1; i >= 0; i-- {
+			if number >= be.Intermediates[i] {
+				return be.Intermediates[i]
+			}
+		}
+	}
+	return number - (number % be.PreviousForkSpec.EpochLength)
+}
+
+func (be BoundaryEpochs) PreviousEpochBlockNumber(currentEpochBlockNumber uint64) uint64 {
+	if currentEpochBlockNumber == 0 {
+		return 0
+	}
+	if currentEpochBlockNumber <= be.PrevLast {
+		return currentEpochBlockNumber - be.PreviousForkSpec.EpochLength
+	}
+
+	for i, mid := range be.Intermediates {
+		if currentEpochBlockNumber == mid {
+			if i == 0 {
+				return be.PrevLast
+			}
+			return be.Intermediates[i-1]
+		}
+	}
+
+	if currentEpochBlockNumber == be.CurrentFirst {
+		if len(be.Intermediates) == 0 {
+			return be.PrevLast
+		}
+		return be.Intermediates[len(be.Intermediates)-1]
+	}
+
+	return currentEpochBlockNumber - be.CurrentForkSpec.EpochLength
+}
+
+func FindTargetForkSpec(forkSpecs []*ForkSpec, height uint64, timestamp uint64) (*ForkSpec, *ForkSpec, error) {
+	reversed := make([]*ForkSpec, len(forkSpecs))
+	for i, spec := range forkSpecs {
+		reversed[len(forkSpecs)-i-1] = spec
+	}
+
+	getPrev := func(current *ForkSpec, i int) *ForkSpec {
+		if i == len(reversed)-1 {
+			return current
+		}
+		return reversed[i+1]
+	}
+
+	for i, spec := range reversed {
+		if x, ok := spec.GetHeightOrTimestamp().(*ForkSpec_Height); ok {
+			if x.Height <= height {
+				return spec, getPrev(spec, i), nil
+			}
+		} else {
+			if spec.GetTimestamp() <= timestamp {
+				return spec, getPrev(spec, i), nil
+			}
+		}
+	}
+	return nil, nil, fmt.Errorf("no fork spec found height=%d, timestmp=%d", height, timestamp)
+}
+
+var boundaryHeightCache = make(map[uint64]uint64)
+
+func GetBoundaryHeight(headerFn getHeaderFn, currentHeight uint64, currentForkSpec ForkSpec) (*BoundaryHeight, error) {
+	logger := log.GetLogger()
+	boundaryHeight := uint64(0)
+	if condition, ok := currentForkSpec.GetHeightOrTimestamp().(*ForkSpec_Height); ok {
+		boundaryHeight = condition.Height
+	} else {
+		ts := currentForkSpec.GetTimestamp()
+		if v, ok := boundaryHeightCache[ts]; ok {
+			boundaryHeight = v
+		} else {
+			logger.Debug("seek fork height", "currentHeight", currentHeight, "ts", ts)
+			for i := int64(currentHeight); i >= 0; i-- {
```

Comment:
Optional optimization: Consider estimating heights near hardfork points using ForkSpec's block generation time to reduce header query overhead. This isn't a required fix but could improve performance for post-hardfork requests.

Reply:
It will be supported after the next version release. There is a grace period until the next Maxwell HF.